handlers/oauth/database.go

```diff
@@ -0,0 +1,52 @@
+package oauth
+
+import (
+	"errors"
+
+	"github.com/speedrun-website/leaderboard-backend/model"
+	"gorm.io/gorm"
+)
+
+type gormContainer struct {
+	DB *gorm.DB
+}
+
+type OauthStore interface {
+	GetUserByTwitterID(string) (*model.User, error)
+	CreateUser(model.User) (*model.User, error)
+}
+
+var Oauth OauthStore
+
+//initGormContainer initializes the oauth store with a gorm container
+func InitGormContainer(db *gorm.DB) {
+	Oauth = &gormContainer{
+		DB: db,
+	}
+}
+
+//GetUserByTwitterID fetches a user by their twitter ID
+func (s gormContainer) GetUserByTwitterID(twitterID string) (*model.User, error) {
+	var maybeExistingUser model.User
+	result := s.DB.Model(&model.User{}).Where("twitter_id = ?", twitterID).First(&maybeExistingUser)
+
+	// No error means we found a user
+	if result.Error == nil {
+		return &maybeExistingUser, nil
+	}
+	isNotFoundError := errors.Is(result.Error, gorm.ErrRecordNotFound)
+
+	// We got a real error
+	if !isNotFoundError {
+		return nil, result.Error
+	}
+	return nil, nil
```

Comment:
Generally if the function did was not successful, it should return some kind of error; returning `nil` for the error generally implies success. Why do we want to return `nil` in this case? Maybe there's a compromise that still returns some kind of error here.

Reply:
Im not entirely sure what our policy should be here. This is technically not an error as not find a user is perfectly expected. I think the only reason gorm errors is because it attempts to update a pointer and theres no easy to say that 'failed'.

Reply:
I'd say that in this function that we call `GetUserByTwitterID`, I'd expect _to_ find a user. And since we couldn't find a user with that Twitter ID, you could reasonably throw an error. Maybe we can have a `UserNotFoundError` which we return in both cases where Gorm errors (i.e. `gorm.ErrRecordNotFound`) and when a record simply doesn't exist.